native/service/governance/neo3_state_manager/utils.go

```diff
@@ -0,0 +1,263 @@
+package neo3_state_manager
+
+import (
+	"fmt"
+	"github.com/polynetwork/poly/common"
+	cstates "github.com/polynetwork/poly/core/states"
+	"github.com/polynetwork/poly/native"
+	"github.com/polynetwork/poly/native/event"
+	"github.com/polynetwork/poly/native/service/utils"
+)
+
+func SerializeStringArray(data []string) []byte {
+	sink := common.NewZeroCopySink(nil)
+	// serialize
+	sink.WriteVarUint(uint64(len(data)))
+	for _, v := range data {
+		sink.WriteString(v)
+	}
+	return sink.Bytes()
+}
+
+func DeserializeStringArray(data []byte) ([]string, error) {
+	source := common.NewZeroCopySource(data)
+	n, eof := source.NextVarUint()
+	if eof {
+		return nil, fmt.Errorf("source.NextVarUint error")
+	}
+	result := make([]string, 0, n)
+	for i := 0; uint64(i) < n; i++ {
+		ss, eof := source.NextString()
+		if eof {
+			return nil, fmt.Errorf("source.NextString error")
+		}
+		result = append(result, ss)
+	}
+	return result, nil
+}
+
+func getStateValidators(native *native.NativeService) ([]string, error) {
+	contract := utils.Neo3StateManagerContractAddress
+	svStore, err := native.GetCacheDB().Get(utils.ConcatKey(contract, []byte(STATE_VALIDATOR)))
+	if err != nil {
+		return nil, fmt.Errorf("getStateValidator, get stateValidatorListParamStore error: %v", err)
+	}
+	if svStore == nil {
+		return []string{}, nil
+	}
+	svBytes, err := cstates.GetValueFromRawStorageItem(svStore)
+	if err != nil {
+		return nil, fmt.Errorf("getStateValidator, deserialize from raw storage item error: %v", err)
+	}
+	svs, err := DeserializeStringArray(svBytes)
+	if err != nil {
+		return nil, fmt.Errorf("getStateValidator, convert to UInt160 array error: %v", err)
+	}
+	return svs, nil
+}
+
+func putStateValidators(native *native.NativeService, stateValidators []string) error {
+	contract := utils.Neo3StateManagerContractAddress
+	// get current stored value
+	oldSVs, err := getStateValidators(native)
+	if err != nil {
+		return fmt.Errorf("putStateValidator, get old state validators error: %v", err)
+	}
+	// use a map to filter old svs
+	mm := make(map[string]string)
```

Comment:
why use map remove the duplicate key, the golang map is out-of-order, so if you use map, every node will run different result, you also store the result in the db, the state root maybe not same.

Reply:
udpated